File: bas/bas.py
import ast
import astor
import dis
#import meta
#import codegen


def ast_to_mermaid(code):
    tree = ast.parse(code)  # gnérer l'arbre AST
    graph = []    # # Initialiser le Mermaid
    
    def visit(node, parent=None):
        node_id = str(id(node))
        label = type(node).__name__ # Utiliser le nom du type de node comme label
        graph.append(f'{node_id}["{label}"]')
        if parent:
            graph.append(f"{parent} --> {node_id}")
        for child in ast.iter_child_nodes(node):
            visit(child, node_id)

    visit(tree)
    return "flowchart TD\\n"+"\\n".join(graph)

code = '''x = 10
y = 12
if x > 5:
    print("grand")
else:
    print("petit")'''
#print(ast_to_mermaid(code).replace("\\n", "\n"))

class mynode(object):
    def __init__(self, node, node_from = None, node_to = None):
        self.node_type = type(node).__name__
        
    def __str__(self):
        if isinstance(node, ast.Assign):
            return "["+ node.targets[0].id + "=" + node.value.n + "]"
        elif isinstance(node, ast.If):
            return "{" + node.targets[0].id + "}"
        

    def set_to(self, id_to):
        self.node_to = id_to
    def set_from(self, id_from):
        self.node_from = id_from

tree = ast.parse(code)
#print(ast.dump(tree, indent=2))  # Affichage + compact


for child in ast.iter_child_nodes(tree):
    print("-----------------------------------")
    if isinstance(child, ast.Assign):
        #print(type(child).__name__)
        #print(child.targets[0].id, "=", child.value.n)  # Affichage de la cible de l'assignation
        print((str((astor.to_source(child)))).strip())
        #print(str(id(child)))
    elif isinstance(child, ast.If):
        #print(type(child).__name__)
        print((str(astor.to_source(child))).strip().splitlines())
        #print("Condition:", ast.dump(child.test))
        #print("Corps: A CONSTRUIRE AVEC...", ast.dump(child))
    elif isinstance(child, ast.Expr):
        print(type(child).__name__)
        print("Expression:", ast.dump(child.value))
    else:
        print("\n AUTRE : ",type(child).__name__)

    print('line',child.lineno, 'col',child.col_offset)  # pour info


##### SOURCE
'''https://docs.python.org/3/library/ast.html
ast.iter_fields(node)
Yield a tuple of (fieldname, value) for each field in node._fields that is present on node.

ast.iter_child_nodes(node)
Yield all direct child nodes of node, that is, all fields that are nodes and all items of fields that are lists of nodes.

ast.walk(node)
Recursively yield all descendant nodes in the tree starting at node (including node itself), in no specified order. This is useful if you only want to modify nodes in place and don’t care about the context.

class ast.NodeVisitor
A node visitor base class that walks the abstract syntax tree and calls a visitor function for every node found. This function may return a value which is forwarded by the visit() method.

This class is meant to be subclassed, with the subclass adding visitor methods.

visit(node)
Visit a node. The default implementation calls the method called self.visit_classname where classname is the name of the node class, or generic_visit() if that method doesn’t exist.

generic_visit(node)
This visitor calls visit() on all children of the node.

'''

File: bas/bas.js
let editor;

// tentative de fix
document.addEventListener('DOMContentLoaded', async function() {
    mermaid.initialize({
        startOnLoad: true,
        theme: "default"
    });

// la variable JavaScript mermaidCode contient le texte Mermaid généré par Pyodide

    // CodeMirror editor
    // const ...
    editor = CodeMirror.fromTextArea(document.getElementById('code-input'), {
        mode: 'python',
        theme: 'dracula',
        lineNumbers: true,
        indentUnit: 4,
        matchBrackets: true
    });

    // un code python par défaut
    editor.setValue(`x = 10
if x > 5:
    print("grand")
else:
    print("petit")`);
    
    await main(); //(editor)
});

async function main() { //(editor)
    
    // Charger Pyodide
    let pyodide = await loadPyodide();

    //code python à recevoir depuis le textarea de codemirror
    const codePython = editor.getValue();

    // Analyse du code Python et génération du graphe Mermaid
    await pyodide.runPythonAsync(`
import ast
import astor
import graphviz

def ast_to_mermaid(codeinput):
    tree = ast.parse(codeinput)  # gnérer l'arbre AST
    graph = []    # # Initialiser le Mermaid
    parents_dict = {}
    
    def visit(node, parent=None):
        node_id = str(id(node))
        label = type(node).__name__ # Utiliser le nom du type de node comme label

        if isinstance(node, ast.Assign):
            label = (str((astor.to_source(node)))).strip()
            graph.append(f'{node_id}["{label}"]')
            if parent is not None:
                parents_dict[node_id] = parent #[node] ou [node_id]

        if isinstance(node, ast.If):
            n = len(str(astor.to_source(node)).strip().splitlines()[0])
            label = "{"+(str(astor.to_source(node))).strip().splitlines()[0][3:n-1]+"}"
            graph.append(f'{node_id}{label}')
            if parent is not None:
                parents_dict[node_id] = parent #[node] ou [node_id]

        #if parent:
            #graph.append(f"{parent} --> {node_id}")

        for child in ast.iter_child_nodes(node):
            visit(child, node_id)

    visit(tree)
    return "flowchart TD\\n"+"\\n".join(graph[:])
    `);

    //alternative
    /* 
    let mermaidGraph = pyodide.globals.get("ast_to_mermaid")(codePython);
    document.getElementById("mermaid").innerHTML = mermaidGraph;
    // Rendu Mermaid.js
    mermaid.init(undefined, document.querySelectorAll(".mermaid"));
*/
    let ast_to_mermaid = pyodide.globals.get("ast_to_mermaid");
    let mermaidGraph = ast_to_mermaid(codePython);

    // injecter le graphe Mermaid dans le conteneur HTML existant
    await afficherFlowchart(mermaidGraph);

}

// main();

// Fonction pour injecter le graphe Mermaid dans le conteneur HTML existant
async function afficherFlowchart(mermaidCode) {
    const container = document.getElementById('flowchart');
    // Injection du contenu Mermaid dans le conteneur
    container.innerHTML = `<pre class="mermaid">${mermaidCode}</pre>`;
    // Déclenchement du rendu Mermaid.js
    await mermaid.run({
        nodes: [container.querySelector('.mermaid')],
    });
}